paimon-core/src/main/java/org/apache/paimon/operation/AbstractFileStoreScan.java

```diff
@@ -195,6 +196,70 @@ public FileStoreScan withManifestCacheFilter(ManifestCacheFilter manifestFilter)
+                                        partitionConverter.getArity() > 0
+                                                ? "partition "
+                                                        + FileStorePathFactory.getPartitionComputer(
+                                                                        partitionConverter
+                                                                                .rowType(),
+                                                                        FileStorePathFactory
+                                                                                .PARTITION_DEFAULT_NAME
+                                                                                .defaultValue())
+                                                                .generatePartValues(
+                                                                        file.partition())
+                                                : "table";
+                                throw new RuntimeException(
+                                        String.format(
+                                                "Try to write %s with a new bucket num %d, but the previous bucket num is %d. "
+                                                        + "Please switch to batch mode, and perform INSERT OVERWRITE to rescale current data layout first.",
+                                                partInfo, numOfBuckets, file.totalBuckets()));
+                            }
+
+                            // bucket filter should not be applied along with partition filter
+                            // because the specifiedBucket is computed against the current
+                            // numOfBuckets
+                            // however entry.bucket() was computed against the old numOfBuckets
+                            // and thus the filtered manifest entries might be empty
+                            // which renders the bucket check invalid
+                            if (filterByBucket(file)
+                                    && filterByBucketSelector(file)
+                                    && filterByLevel(file)) {
+                                return true;
+                            }
+                            return false;
+                        });
+        final Long readSnapshotId = planResult.getLeft();
+        final List<ManifestEntry> files = planResult.getRight();
+
+        return new Plan() {
+            @Nullable
+            @Override
+            public Long snapshotId() {
+                return readSnapshotId;
+            }
+
+            @Override
+            public List<ManifestEntry> files() {
+                return files;
+            }
+        };
+    }
+
+    private <T extends AbstractManifestEntry> Pair<Long, List<T>> doPlan(
+            Function<List<ManifestFileMeta>, List<T>> processor,
+            Function<T, Boolean> postFilterProcessor) {
```

Comment:
Maybe we don't need this one, postFilter can be applied to `SimpleManifestEntry` too.
If there is level filter, throw exception.